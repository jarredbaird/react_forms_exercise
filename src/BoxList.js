import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css";

const BoxList = (props) => {
  const INITIAL_STATE = [];

  const [boxes, setBoxes] = useState(INITIAL_STATE);

  const addBox = (newBox) => {
    setBoxes((boxes) => [...boxes, { ...newBox, id: uuid() }]);
  };
  const removeBox = (boxId) => {
    setBoxes((boxes) => boxes.filter((box) => box.id !== boxId));
  };
  return (
    <>
      <NewBoxForm addBox={addBox} />
      <div className="BoxItem">
        <div>
          {boxes.map(({ id, color, height, width }) => {
            return (
              <Box
                key={id}
                id={id}
                color={color}
                height={height}
                width={width}
                removeBox={removeBox}
              />
            );
          })}
        </div>
      </div>
    </>
  );
};

export default BoxList;
