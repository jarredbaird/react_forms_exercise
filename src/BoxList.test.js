import React from "react";
import { render, fireEvent, queryByAttribute } from "@testing-library/react";
import BoxList from "./BoxList";

test("BoxList renders", () => {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("creates a new box", () => {
  const { queryByTestId, queryByText } = render(<BoxList />);

  // no boxes should be here yet
  expect(queryByTestId("box")).not.toBeInTheDocument();

  const colorInput = queryByTestId("color");
  const widthInput = queryByTestId("width");
  const heightInput = queryByTestId("height");
  const submitButton = queryByText("Add the best box ever!");

  fireEvent.change(colorInput, { target: { value: "blue" } });
  fireEvent.change(widthInput, { target: { value: "50" } });
  fireEvent.change(heightInput, { target: { value: "60" } });
  fireEvent.click(submitButton);

  // box created?!
  expect(queryByTestId("box")).toBeInTheDocument();
});

it("removes a single box", () => {
  const { queryByTestId, queryByText } = render(<BoxList />);

  // no boxes should be here yet
  expect(queryByTestId("box")).not.toBeInTheDocument();

  const colorInput = queryByTestId("color");
  const widthInput = queryByTestId("width");
  const heightInput = queryByTestId("height");
  const submitButton = queryByText("Add the best box ever!");

  fireEvent.change(colorInput, { target: { value: "blue" } });
  fireEvent.change(widthInput, { target: { value: "50" } });
  fireEvent.change(heightInput, { target: { value: "60" } });
  fireEvent.click(submitButton);

  // box created?!
  expect(queryByTestId("box")).toBeInTheDocument();

  const removeBtn = queryByText("X");
  fireEvent.click(removeBtn);
  expect(queryByTestId("box")).not.toBeInTheDocument();
});
