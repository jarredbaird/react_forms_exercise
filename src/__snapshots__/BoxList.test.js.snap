// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="name"
    >
      Color: 
    </label>
    <input
      data-testid="color"
      id="color"
      name="color"
      placeholder="Box Color"
      type="text"
      value=""
    />
  </form>
  <p />
  <form>
    <label
      for="name"
    >
      Width: 
    </label>
    <input
      data-testid="width"
      id="width"
      name="width"
      placeholder="Box Width"
      type="text"
      value=""
    />
  </form>
  <p />
  <form>
    <label
      for="name"
    >
      Height: 
    </label>
    <input
      data-testid="height"
      id="height"
      name="height"
      placeholder="Box Height"
      type="text"
      value=""
    />
    <button>
      Add the best box ever!
    </button>
  </form>
  <p />
  <div
    class="BoxItem"
  >
    <div />
  </div>
</DocumentFragment>
`;
